pkg/controller/meta_pod_controller.go

```diff
@@ -0,0 +1,175 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/risingwavelabs/risingwave-operator/pkg/consts"
+	pb "github.com/risingwavelabs/risingwave-operator/pkg/controller/proto"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+// Tutorial: https://grpc.io/docs/languages/go/quickstart/
+// Generate proto files via make proto
+
+// MetaPodController reconciles meta pods object.
+type MetaPodController struct {
+	client.Client
+}
+
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;update;patch
+
+// metaLeaderStatus sends a MetaMember request to a meta node at ip:port, determining if the node is a leader.
+func (mpc *MetaPodController) metaLeaderStatus(ctx context.Context, host string, port uint) string {
+	log := log.FromContext(ctx)
+
+	addr := fmt.Sprintf("%s:%v", host, port)
+
+	for i := 0; i < 5; i++ {
+		time.Sleep(time.Duration(i*10) * time.Millisecond)
+		conn, err := grpc.Dial(addr, grpc.WithTransportCredentials(insecure.NewCredentials()))
+		if err != nil {
+			log.Error(err, "Unable to connect to meta pod. Retrying...")
+			continue
+		}
+		defer conn.Close()
+		c := pb.NewMetaMemberServiceClient(conn)
+
+		resp, err := c.Members(ctx, &pb.MembersRequest{})
+		if err != nil {
+			log.Error(err, "Sending MembersRequest failed")
+			continue
+		}
+
+		for _, member := range resp.Members {
+			if member.IsLeader && host == member.Address.Host && port == uint(member.Address.Port) {
+				return consts.MetaRoleLeader
+			}
+		}
+		return consts.MetaRoleFollower
+	}
+	return consts.MetaRoleUnknown
+}
+
+// getMetaPort returns the service port of the pod.
+func getMetaPort(pod *corev1.Pod) (uint, error) {
+	for _, container := range pod.Spec.Containers {
+		if container.Name == "meta" {
+			for _, containerPort := range container.Ports {
+				if containerPort.Name == "service" {
```

Comment:
nit: there's a constant to use
```suggestion
				if containerPort.Name == consts.PortService {
```